File: src/experimental/workflows/agent.ts
/**
 * Copyright 2025 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BeeAgent } from "@/agents/bee/agent.js";
import { Workflow, WorkflowRunOptions } from "@/experimental/workflows/workflow.js";
import { AssistantMessage, Message } from "@/backend/message.js";
import { AnyTool } from "@/tools/base.js";
import { BaseMemory, ReadOnlyMemory } from "@/memory/base.js";
import { z } from "zod";
import { UnconstrainedMemory } from "@/memory/unconstrainedMemory.js";
import { BaseAgent } from "@/agents/base.js";
import {
  BeeAgentExecutionConfig,
  BeeRunInput,
  BeeRunOptions,
  BeeRunOutput,
} from "@/agents/bee/types.js";
import { isFunction, randomString } from "remeda";
import { ChatModel } from "@/backend/chat.js";

type AgentInstance = BaseAgent<BeeRunInput, BeeRunOutput, BeeRunOptions>;
type AgentFactory = (memory: ReadOnlyMemory) => AgentInstance | Promise<AgentInstance>;
interface AgentFactoryInput {
  name: string;
  llm: ChatModel;
  instructions?: string;
  tools?: AnyTool[];
  execution?: BeeAgentExecutionConfig;
}

export class AgentWorkflow {
  protected readonly workflow;

  static readonly schema = z.object({
    messages: z.array(z.instanceof(Message)).min(1),

    finalAnswer: z.string().optional(),
    newMessages: z.array(z.instanceof(Message)).default([]),
  });

  constructor(name = "AgentWorkflow") {
    this.workflow = new Workflow({
      name,
      schema: AgentWorkflow.schema,
      outputSchema: AgentWorkflow.schema.required(),
    });
  }

  run(messages: Message[], options: WorkflowRunOptions<string> = {}) {
    return this.workflow.run(
      {
        messages,
      },
      options,
    );
  }

  addAgent(agent: AgentFactory | AgentFactoryInput): this;
  addAgent(agent: AgentInstance): Promise<this>;
  addAgent(agent: AgentInstance | AgentFactory | AgentFactoryInput): this | Promise<this> {
    if (agent instanceof BaseAgent) {
      return agent.clone().then((clone) => {
        const factory: AgentFactory = (memory) => {
          clone.memory = memory;
          return clone;
        };
        return this._add(clone.meta.name, factory);
      });
    }

    const name = agent.name || `Agent${randomString(4)}`;
    return this._add(name, isFunction(agent) ? agent : this._createFactory(agent));
  }

  delAgent(name: string) {
    return this.workflow.delStep(name);
  }

  protected _createFactory(input: AgentFactoryInput): AgentFactory {
    return (memory: BaseMemory) =>
      new BeeAgent({
        llm: input.llm,
        tools: input.tools ?? [],
        memory,
        meta: {
          name: input.name,
          description: input.instructions ?? "",
        },
        execution: input.execution,
        ...(input.instructions && {
          templates: {
            system: (template) =>
              template.fork((config) => {
                config.defaults.instructions = input.instructions || config.defaults.instructions;
              }),
          },
        }),
      });
  }

  protected _add(name: string, factory: AgentFactory) {
    this.workflow.addStep(name, async (state, ctx) => {
      const memory = new UnconstrainedMemory();
      await memory.addMany([...state.messages, ...state.newMessages]);

      const agent = await factory(memory.asReadOnly());
      const { result } = await agent.run({ prompt: null }, { signal: ctx.signal });

      return {
        update: {
          finalAnswer: result.text,
          newMessages: state.newMessages.concat(
            new AssistantMessage(
              [`Assistant Name: ${name}`, `Assistant Response: ${result.text}`].join("\n"),
            ),
          ),
        },
      };
    });
    return this;
  }
}

File: python/beeai_framework/workflows/agent.py
# Copyright 2025 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import asyncio
import random
import string
from collections.abc import Callable

from pydantic import BaseModel, ConfigDict, Field, InstanceOf

from beeai_framework.agents.base import BaseAgent, BaseMemory
from beeai_framework.agents.bee import BeeAgent
from beeai_framework.agents.types import (
    AgentMeta,
    BeeAgentExecutionConfig,
    BeeInput,
    BeeRunInput,
    BeeRunOutput,
)
from beeai_framework.backend.chat import ChatModel
from beeai_framework.backend.message import AssistantMessage, Message
from beeai_framework.memory import ReadOnlyMemory, UnconstrainedMemory
from beeai_framework.tools.tool import Tool
from beeai_framework.workflows.workflow import Workflow, WorkflowRun

AgentFactory = Callable[[ReadOnlyMemory], BaseAgent | asyncio.Future[BaseAgent]]


class AgentFactoryInput(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)
    name: str
    llm: ChatModel
    instructions: str | None = None
    tools: list[InstanceOf[Tool]] | None = None
    execution: BeeAgentExecutionConfig | None = None


class Schema(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)
    messages: list[Message] = Field(min_length=1)
    final_answer: str | None = None
    new_messages: list[Message] = []


class AgentWorkflow:
    def __init__(self, name: str = "AgentWorkflow") -> "AgentWorkflow":
        self.workflow = Workflow(name=name, schema=Schema)

    async def run(self, messages: list[Message]) -> WorkflowRun:
        return await self.workflow.run(Schema(messages=messages))

    def del_agent(self, name: str) -> "AgentWorkflow":
        return self.workflow.delete_step(name)

    def add_agent(
        self, agent: BaseAgent | Callable[[ReadOnlyMemory], BaseAgent | asyncio.Future[BaseAgent]] | AgentFactoryInput
    ) -> "AgentWorkflow":
        if isinstance(agent, BaseAgent):

            def factory(memory: ReadOnlyMemory) -> AgentFactory:
                agent.memory = memory
                return agent

            return self._add(agent.meta.name, factory)

        random_string = "".join(random.choice(string.ascii_letters) for _ in range(4))
        name = agent.name if not callable(agent) else f"Agent{random_string}"
        return self._add(name, agent if callable(agent) else self._create_factory(agent))

    def _create_factory(self, input: AgentFactoryInput) -> AgentFactory:
        def factory(memory: BaseMemory) -> BeeAgent:
            return BeeAgent(
                bee_input=BeeInput(
                    llm=input.llm,
                    tools=input.tools or [],
                    memory=memory,
                    # template TODO
                    meta=AgentMeta(name=input.name, description=input.instructions or "", tools=[]),
                    execution=input.execution,
                )
            )

        return factory

    def _add(self, name: str, factory: AgentFactory) -> "AgentWorkflow":
        async def step(state: Schema) -> None:
            memory = UnconstrainedMemory()
            for message in state.messages + state.new_messages:
                await memory.add(message)

            agent: BaseAgent = factory(memory.as_read_only())
            run_output: BeeRunOutput = await agent.run(run_input=BeeRunInput())
            state.final_answer = run_output.result.text
            state.new_messages.append(
                AssistantMessage(f"Assistant Name: {name}\nAssistant Response: {run_output.result.text}")
            )

        self.workflow.add_step(name, step)
        return self
